1.Data load karo (poora dataset, jaisa Day 27 mein seekha):

In [1]:
import pandas as pd

df = pd.read_csv(
    '../data/raw/online_retail_II.csv',
    dtype={'Invoice': str, 'StockCode': str},
    parse_dates=['InvoiceDate']
)
print(df.shape)

(1067371, 8)


2.head() aur tail() — pehli/aakhri rows dekhna:

In [2]:
print(df.head())        # default 5 rows
print(df.head(10))      # custom count
print(df.tail())        # aakhri 5 rows
print(df.sample(5))     # random 5 rows — zyada useful hai kabhi kabhi (head sirf shuru dikhata hai, bias ho sakta hai)

  Invoice StockCode                          Description  Quantity  \
0  489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1  489434    79323P                   PINK CHERRY LIGHTS        12   
2  489434    79323W                  WHITE CHERRY LIGHTS        12   
3  489434     22041         RECORD FRAME 7" SINGLE SIZE         48   
4  489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   

          InvoiceDate  Price  Customer ID         Country  
0 2009-12-01 07:45:00   6.95      13085.0  United Kingdom  
1 2009-12-01 07:45:00   6.75      13085.0  United Kingdom  
2 2009-12-01 07:45:00   6.75      13085.0  United Kingdom  
3 2009-12-01 07:45:00   2.10      13085.0  United Kingdom  
4 2009-12-01 07:45:00   1.25      13085.0  United Kingdom  
  Invoice StockCode                          Description  Quantity  \
0  489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1  489434    79323P                   PINK CHERRY LIGHTS        12   
2  489434

3.info() — poori DataFrame ka summary (data types, non-null counts, memory):

In [3]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1067371 entries, 0 to 1067370
Data columns (total 8 columns):
 #   Column       Non-Null Count    Dtype         
---  ------       --------------    -----         
 0   Invoice      1067371 non-null  str           
 1   StockCode    1067371 non-null  str           
 2   Description  1062989 non-null  str           
 3   Quantity     1067371 non-null  int64         
 4   InvoiceDate  1067371 non-null  datetime64[us]
 5   Price        1067371 non-null  float64       
 6   Customer ID  824364 non-null   float64       
 7   Country      1067371 non-null  str           
dtypes: datetime64[us](1), float64(2), int64(1), str(4)
memory usage: 65.1 MB


yeh dekh kar samjho:


1.Har column ka Dtype (int64, object, float64, datetime64)

2.Non-Null Count — kitni values missing nahi hain (isse Customer ID ka gap turant dikh jaayega)

3.memory usage — poora dataset kitni RAM le raha hai

4.describe() — numerical columns ka statistical summary:

In [4]:
print(df.describe())

           Quantity                 InvoiceDate         Price    Customer ID
count  1.067371e+06                     1067371  1.067371e+06  824364.000000
mean   9.938898e+00  2011-01-02 21:13:55.394029  4.649388e+00   15324.638504
min   -8.099500e+04         2009-12-01 07:45:00 -5.359436e+04   12346.000000
25%    1.000000e+00         2010-07-09 09:46:00  1.250000e+00   13975.000000
50%    3.000000e+00         2010-12-07 15:28:00  2.100000e+00   15255.000000
75%    1.000000e+01         2011-07-22 10:23:00  4.150000e+00   16797.000000
max    8.099500e+04         2011-12-09 12:50:00  3.897000e+04   18287.000000
std    1.727058e+02                         NaN  1.235531e+02    1697.464450


Note = count, mean, std, min, 25%, 50% (median), 75%, max — yeh sab Day 25 mein NumPy se manually nikale the, describe() sab ek saath de deta hai.

5.describe() categorical (text) columns ke liye bhi:

In [5]:
print(df.describe(include='object'))

C:\Users\rajat\AppData\Local\Temp\ipykernel_8792\3781773561.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  print(df.describe(include='object'))


        Invoice StockCode                         Description         Country
count   1067371   1067371                             1062989         1067371
unique    53628      5305                                5698              43
top      537434    85123A  WHITE HANGING HEART T-LIGHT HOLDER  United Kingdom
freq       1350      5829                                5918          981330


Note = Yeh count, unique, top (sabse common value), freq (uski frequency) dikhayega — jaise Country column mein kaunsa country sabse zyada hai.

6.describe() poore dataset (numeric + categorical dono) ke liye:

In [6]:
print(df.describe(include='all'))

        Invoice StockCode                         Description      Quantity  \
count   1067371   1067371                             1062989  1.067371e+06   
unique    53628      5305                                5698           NaN   
top      537434    85123A  WHITE HANGING HEART T-LIGHT HOLDER           NaN   
freq       1350      5829                                5918           NaN   
mean        NaN       NaN                                 NaN  9.938898e+00   
min         NaN       NaN                                 NaN -8.099500e+04   
25%         NaN       NaN                                 NaN  1.000000e+00   
50%         NaN       NaN                                 NaN  3.000000e+00   
75%         NaN       NaN                                 NaN  1.000000e+01   
max         NaN       NaN                                 NaN  8.099500e+04   
std         NaN       NaN                                 NaN  1.727058e+02   

                       InvoiceDate         Price   

7.Individual column ki jhalak — unique(), nunique(), value_counts():

In [7]:
print("Unique countries:", df['Country'].nunique())
print(df['Country'].unique())          # saari unique values ki list
print(df['Country'].value_counts())    # har value ka count, sorted
print(df['Country'].value_counts(normalize=True) * 100)  # percentage mein

Unique countries: 43
<StringArray>
[      'United Kingdom',               'France',                  'USA',
              'Belgium',            'Australia',                 'EIRE',
              'Germany',             'Portugal',                'Japan',
              'Denmark',              'Nigeria',          'Netherlands',
               'Poland',                'Spain',      'Channel Islands',
                'Italy',               'Cyprus',               'Greece',
               'Norway',              'Austria',               'Sweden',
 'United Arab Emirates',              'Finland',          'Switzerland',
          'Unspecified',                'Malta',              'Bahrain',
                  'RSA',              'Bermuda',            'Hong Kong',
            'Singapore',             'Thailand',               'Israel',
            'Lithuania',          'West Indies',              'Lebanon',
                'Korea',               'Brazil',               'Canada',
              'I

8.dtypes aur columns dobara check karna (Day 26 mein bhi dekha tha):

In [8]:
print(df.dtypes)
print(df.columns.tolist())

Invoice                   str
StockCode                 str
Description               str
Quantity                int64
InvoiceDate    datetime64[us]
Price                 float64
Customer ID           float64
Country                   str
dtype: object
['Invoice', 'StockCode', 'Description', 'Quantity', 'InvoiceDate', 'Price', 'Customer ID', 'Country']


9.df.isnull().sum() — quick missing value check (Day 32 mein detail se karenge, aaj sirf overview):

In [10]:
print(df.isnull().sum())

Invoice             0
StockCode           0
Description      4382
Quantity            0
InvoiceDate         0
Price               0
Customer ID    243007
Country             0
dtype: int64


Practice questions

1.df['StockCode'].value_counts().head(10) se pata karo sabse zyada baar kaunsa StockCode appear hota hai (sabse popular product).

Ans: is output me sabse top par 85123A stcokcode hai, jo 5829 baar appear hua hai. iska matlab yeh hai ki yeh dataset ka sabse popular product hai.

In [13]:
df['StockCode'].value_counts().head(10)

StockCode
85123A    5829
22423     4424
85099B    4216
21212     3318
20725     3259
84879     2960
47566     2768
21232     2747
22197     2549
22383     2540
Name: count, dtype: int64

2.df.describe() ke output se manually batao: Quantity column ka minimum negative kyun hai? (Hint: Day 10 mein dekha tha — returns/cancellations se related hai).

Ans: Quantity ka minimum value -8.0099500+04 (yaani -80995) hai. jab koi customer order cancel karta hai ya product return karta hai , to system me us transaction ke negative ke sath record kiya jata hai.

In [14]:
df.describe()

,Quantity,InvoiceDate,Price,Customer ID
count,1.067371e+06,1067371,1.067371e+06,824364.000000
mean,9.938898e+00,2011-01-02 21:13:55.394029,4.649388e+00,15324.638504
min,-8.099500e+04,2009-12-01 07:45:00,-5.359436e+04,12346.000000
25%,1.000000e+00,2010-07-09 09:46:00,1.250000e+00,13975.000000
50%,3.000000e+00,2010-12-07 15:28:00,2.100000e+00,15255.000000
75%,1.000000e+01,2011-07-22 10:23:00,4.150000e+00,16797.000000
max,8.099500e+04,2011-12-09 12:50:00,3.897000e+04,18287.000000
std,1.727058e+02,NaN,1.235531e+02,1697.464450
